vm/systemSmartContracts/governance.go

```diff
@@ -812,98 +1101,129 @@ func (g *governanceContract) numOfStakedNodes(address []byte) (uint32, error) {
-	}
+	return votingPower, nil
+}
 
-	generalProposal.Closed = true
-	err = g.computeEndResults(generalProposal)
-	if err != nil {
-		g.eei.AddReturnMessage("computeEndResults error" + err.Error())
-		return vmcommon.UserError
+// validateInitialWhiteListedAddresses makes basic checks that the provided initial whitelisted
+//  addresses have the correct format
+func (g *governanceContract) validateInitialWhiteListedAddresses(addresses [][]byte) error {
+	if len(addresses) == 0 {
+		log.Debug("0 initial whiteListed addresses provided to the governance contract")
+		return nil
 	}
 
-	err = g.saveGeneralProposal(proposal, generalProposal)
-	if err != nil {
-		g.eei.AddReturnMessage("saveGeneralProposal error" + err.Error())
-		return vmcommon.UserError
+	for _, addr := range addresses {
+		if len(addr) != len(g.ownerAddress) {
+			return fmt.Errorf("invalid address length for %s", string(addr))
+		}
 	}
 
-	for _, voter := range generalProposal.Voters {
-		key := append(proposal, voter...)
-		g.eei.SetStorage(key, nil)
+	return nil
+}
+
+// startEndNonceFromArguments converts the nonce string arguments to uint64
+func (g *governanceContract) startEndNonceFromArguments(argStart []byte, argEnd []byte) (uint64, uint64, error) {
+	startVoteNonce, okConvert := big.NewInt(0).SetString(string(argStart), conversionBase)
+	if !okConvert {
+		return 0, 0, vm.ErrInvalidStartEndVoteNonce
+	}
+	if !startVoteNonce.IsUint64() {
+		return 0, 0, vm.ErrInvalidStartEndVoteNonce
+	}
+	endVoteNonce, okConvert := big.NewInt(0).SetString(string(argEnd), conversionBase)
+	if !okConvert {
+		return 0, 0, vm.ErrInvalidStartEndVoteNonce
+	}
+	if !endVoteNonce.IsUint64() {
+		return 0, 0, vm.ErrInvalidStartEndVoteNonce
+	}
+	currentNonce := g.eei.BlockChainHook().CurrentNonce()
+	if currentNonce > startVoteNonce.Uint64() || startVoteNonce.Uint64() > endVoteNonce.Uint64() {
+		return 0, 0, vm.ErrInvalidStartEndVoteNonce
 	}
 
-	return vmcommon.Ok
+	return startVoteNonce.Uint64(), endVoteNonce.Uint64(), nil
 }
 
+// computeEndResults computes if a proposal has passed or not based on votes accumulated
 func (g *governanceContract) computeEndResults(proposal *GeneralProposal) error {
 	baseConfig, err := g.getConfig()
 	if err != nil {
 		return err
 	}
-	totalVotes := proposal.Yes + proposal.No + proposal.DontCare + proposal.Veto
-	if totalVotes < baseConfig.MinQuorum {
+
+	totalVotes := big.NewInt(0).Add(proposal.Yes, proposal.No)
+	totalVotes.Add(totalVotes, proposal.Veto)
+
+	if totalVotes.Cmp(baseConfig.MinQuorum) == -1 {
 		proposal.Voted = false
 		return nil
 	}
 
-	if proposal.Veto > baseConfig.MinVetoThreshold {
-		proposal.Voted = false
+	if proposal.Yes.Cmp(baseConfig.MinPassThreshold) == 1 && proposal.Yes.Cmp(proposal.No) == 1 {
+		proposal.Voted = true
 		return nil
 	}
 
-	if proposal.Yes > baseConfig.MinPassThreshold {
-		proposal.Voted = true
+	if proposal.Veto.Cmp(baseConfig.MinVetoThreshold) == 1 {
+		proposal.Voted = false
 		return nil
 	}
 
 	return nil
```

Comment:
Yes. If it was vetoed, it's not voted. Maybe the name here it's actually misleading, would renaming it to `Passed` make more sense?

Reply:
The idea was that if we reach this point and return nil, it should have been because the proposal did not raised enough Yes votes (it was not vetoed, it just didn't reached the 50% + 1 yes votes). So, for being more rigorously, I think we should add here  `proposal.Voted = false` before the return.

Reply:
Ah, yeah. It was weird how github showed your comment and now I realise that we were talking about different lines :). It should be false there, but I will add it to be specific and sure that `Voted` will be false at the end of the function